tests/Microsoft.Identity.Test.Unit/ParallelRequestsTests.cs

```diff
@@ -85,6 +85,138 @@ public async Task ExtraQP()
             Assert.AreEqual(NumberOfRequests, results.Length);
         }
 
+        [TestMethod]
+        public async Task AcquireTokenForClient_ConcurrentTenantRequests_Test()
+        {
+            // Arrange
+            const int NumberOfRequests = 1000;
+
+            // Custom HTTP manager that counts the number of requests
+            ParallelRequestMockHandler httpManager = new();
+
+            var cca = ConfidentialClientApplicationBuilder
+                .Create(TestConstants.ClientId)
+                .WithAuthority("https://login.microsoftonline.com/common")
+                .WithClientSecret(TestConstants.ClientSecret)
+                .WithHttpManager(httpManager)
+                .Build();
+
+            var tasks = new List<Task<AuthenticationResult>>();
+
+            for (int i = 0; i < NumberOfRequests; i++)
+            {
+                tasks.Add(Task.Run(async () =>
+                {
+                    string tid = $"tidtid_{i}";
+                    AuthenticationResult res = await cca.AcquireTokenForClient(TestConstants.s_scope)
+                        .WithTenantId(tid)
+                        .ExecuteAsync()
+                        .ConfigureAwait(false);
+
+                    Assert.IsFalse(
+                        string.IsNullOrEmpty(res.AuthenticationResultMetadata.TokenEndpoint),
+                        "TokenEndpoint is null/empty!"
+                    );
+                    Assert.IsTrue(
+                        res.AuthenticationResultMetadata.TokenEndpoint.Contains(tid),
+                        "TokenEndpoint should contain the tenant ID."
+                    );
+                    Assert.AreEqual($"token_{tid}", res.AccessToken, "Access token did not match the expected value.");
+
+                    return res;
+                }));
+            }
+
+            // Wait for all tasks to complete
+            AuthenticationResult[] results = await Task.WhenAll(tasks).ConfigureAwait(false);
+
+            // Assert the total tasks
+            Assert.AreEqual(NumberOfRequests, results.Length, "Number of AuthenticationResult objects does not match the number of requests.");
+        }
+
+        [TestMethod]
+        public async Task AcquireTokenForClient_PerTenantCaching_Test()
```

Comment:
Ensures that tokens for multiple tenants are stored separately and retrieved from the cache if already requested.